warehouse/migrations/env.py

```diff
@@ -42,37 +42,25 @@ def run_migrations_online():
     In this scenario we need to create an Engine
     and associate a connection with the context.
     """
-    connectable = context.config.attributes.get("connection", None)
+    options = context.config.get_section(context.config.config_ini_section)
+    url = options.pop("url")
+    connectable = create_engine(url, poolclass=pool.NullPool)
 
-    if connectable is None:
-        options = context.config.get_section(context.config.config_ini_section)
-        url = options.pop("url")
-        connectable = create_engine(url, poolclass=pool.NullPool)
+    with connectable.connect() as connection:
+        connection.execute(text("SET statement_timeout = 5000"))
+        connection.execute(text("SET lock_timeout = 4000"))
 
-        with connectable.connect() as connection:
-            connection.execute(text("SET statement_timeout = 5000"))
-            connection.execute(text("SET lock_timeout = 4000"))
-
-            context.configure(
-                connection=connection,
-                target_metadata=db.metadata,
-                compare_server_default=True,
-                transaction_per_migration=True,
-            )
-            with context.begin_transaction():
-                context.run_migrations()
-    else:
         context.configure(
-            connection=connectable,
+            connection=connection,
             target_metadata=db.metadata,
             compare_server_default=True,
             transaction_per_migration=True,
         )
-        context.execute(text("SET statement_timeout = 5000"))
-        context.execute(text("SET lock_timeout = 4000"))
-
         with context.begin_transaction():
+            connection.execute(text("SELECT pg_advisory_lock(hashtext('alembic'))"))
```

Comment:
This is way cleaner for migrations, and I guess we don't strictly need a lock to inspect the status.

Still curious if there was a specific reason for removing the context manager approach though.

Reply:
We removed the ``alembic_lock`` context manager during anti entropy because something about the way it was implemented was causing some of failures. I'm not 100% clear on what exactly it was, but I think the alembic was getting confused by having an already open transaction, due to the alembic lock needing to open one, rather than having alembic open the transactions itself.

This bit could still be in a context manager if we wanted to (it would just need to take the connection object, rather than create it itself like it was previously doing), I think @di did this live while on the call and since it's not being used anywhere else there wasn't a pressing need to abstract it out.

Reply:
@di there needs to be a final commit _after_ the unlock, otherwise it gets rolled at the end of the context.begin_transaction:

```
...
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "UPDATE alembic_version SET version_num='4a0276f260c7' WHERE alembic_version.version_num = '34cccbcab226'", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "[generated in 0.00017s] {}", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "COMMIT", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "BEGIN (implicit)", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "SELECT pg_advisory_unlock(hashtext('alembic'))", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "[generated in 0.00025s] {}", "thread": 281473358847296}
{"logger": "sqlalchemy.engine.Engine", "level": "INFO", "event": "ROLLBACK", "thread": 281473358847296}
...
```

Reply:
Rolling back the unlock doesn't matter

Reply:
Hmmm - you're right @dstufft - but not in the way that I thought. Thanks for pointing that out! I tested it in a couple of shells locally to see how it behaved, and it did what I thought it was supposed to do.